src/udp.rs

```diff
@@ -0,0 +1,377 @@
+pub async fn connect_udp(args: ConnectUdpArgs) -> anyhow::Result<()> {
+    let addrs = args
+        .addr
+        .to_socket_addrs()
+        .context(format!("invalid host string {}", args.addr))?;
+    let secret_key = get_or_create_secret()?;
+    let mut builder = Endpoint::builder().secret_key(secret_key).alpns(vec![]);
+    if let Some(addr) = args.common.magic_ipv4_addr {
+        builder = builder.bind_addr_v4(addr);
+    }
+    if let Some(addr) = args.common.magic_ipv6_addr {
+        builder = builder.bind_addr_v6(addr);
+    }
+    let endpoint = builder.bind().await.context("unable to bind magicsock")?;
+    tracing::info!("udp listening on {:?}", addrs);
+    let socket = Arc::new(UdpSocket::bind(addrs.as_slice()).await?);
+
+    let node_addr = args.ticket.node_addr();
+    let mut buf: Vec<u8> = vec![0u8; 65535];
+    let conns = Arc::new(tokio::sync::Mutex::new(
+        HashMap::<SocketAddr, Connection>::new(),
+    ));
+    loop {
+        tokio::select! {
+            _ = signal::ctrl_c() => {
+                eprintln!("Received CTRL-C, shutting down...");
+                break;
+            }
+            result = socket.recv_from(&mut buf) => {
+                match result {
+                    Ok((size, sock_addr)) => {
+                        // Check if we already have a connection for this socket address
+                        let mut cnns = conns.lock().await;
+                        let connection = match cnns.get_mut(&sock_addr) {
+                            Some(conn) => conn,
+                            None => {
+                                // If we don't have a connection, drop the previous lock to create a new one later on
+                                drop(cnns);
+
+                                // Create a new connection since this address is not in the hashmap
+                                let endpoint = endpoint.clone();
+                                let addr = node_addr.clone();
+                                let handshake = !args.common.is_custom_alpn();
+                                let alpn = args.common.alpn()?;
+
+                                let remote_node_id = addr.node_id;
+                                tracing::info!("creating a connection to be forwarding UDP to {}", remote_node_id);
+
+                                // connect to the node, try only once
+                                let connection = endpoint
+                                    .connect(addr.clone(), &alpn)
+                                    .await
+                                    .context(format!("error connecting to {}", remote_node_id))?;
+                                tracing::info!("connected to {}", remote_node_id);
+
+                                // send the handshake unless we are using a custom alpn
+                                if handshake {
+                                    connection.send_datagram(Bytes::from_static(&dumbpipe::HANDSHAKE))?;
```

Comment:
I guess failing here on `TooLarge` is not a good call, because that is a normal thing that might happen.  You just drop such datagrams and move on I think.